lib/ruby_lsp/requests/support/rubocop_runner.rb

```diff
@@ -109,6 +113,28 @@ def formatted_source
           @options[:stdin]
         end
 
+        sig { params(file: String).returns(RuboCop::ProcessedSource) }
+        def get_processed_source(file)
+          config = @config_store.for_file(file)
+          parser_engine = config.parser_engine
+          return super unless parser_engine == :parser_prism
+
+          processed_source = T.unsafe(::RuboCop::AST::ProcessedSource).new(
+            @options[:stdin],
+            Prism::Translation::Parser::VERSION_3_3,
+            file,
+            parser_engine: parser_engine,
+            prism_result: @parse_result,
+          )
```

Comment:
We could avoid some patching if there was a way to instantiate the `ProcessedSource` with an existing AST and token list. @koic, would you be open to accepting that?

It would allow us to avoid some patching.

Reply:
All features in the Ruby LSP depend only the Prim's parse result, so we keep that in our representation of documents.

For RuboCop, we just [invoke run with the file contents](https://github.com/Shopify/ruby-lsp/blob/049d0e69d2523b7f54a349b4aa1e5e78b4e6c988/lib/ruby_lsp/requests/support/rubocop_runner.rb#L94), which will let it re-parse the file and do everything it needs to do.

The Ruby LSP supports multiple formatters. We hook into RuboCop if it is available, but it's not a dependency. So we can't depend on RuboCop internals for our basic document representations, like using a `ProcessedSource`. All RuboCop things are limited to the runner here.

Now that RuboCop supports the Prism backend, we're trying to find what the easiest way of reusing the existing AST is, so that we can stop parsing documents twice.

The ideal API would be something like what Syntax Tree supports, which allows you to format a given AST. That would not only simplify the code here a lot, it would enable RuboCop to support range formatting for LSPs, which is currently impossible.

That said, the ideal API may require significant refactors, which is why I suggested something a bit simpler that will still minimize the number of patches. If we can instantiate the `ProcessedSource` in a way that allows us to reuse the existing AST and prevents the second parse, that'll already simplify the code quite a bit.

Reply:
But a ProcessedDocument also holds the tokens...

Reply:
The use case this PR aims to address appears reasonable to me. When using a Ruby code with multiple tools (e.g., RuboCop, TypeProf, Steep...), parsing the same source code to an AST in each tool becomes a redundant process. In such cases, it makes sense to reuse the one AST across different tools.

In this PR, the only change to the interface for reusing the AST is the addition of the `prism_result` keyword argument. This means that backward compatibility is maintained, and there is no impact on the existing RuboCop (AST).

In summary, I think this proposal can be accepted by RuboCop and RuboCop AST. Considering this is not a specialized requirement but a backend tool for LSP, it would be more rational to provide a published interface rather than having users write monkey patches. Below, I share one concern and an idea related to it.

## NOTE: Naming and the two ASTs (whitequark AST and Prism AST)

A point of caution is that the current AST RuboCop can analyze is only the whitequark/parser AST interface (`Prism::Translation::Parser`). It cannot process the vanilla Prism AST interface.

Therefore, I'm not sure if `prism_result` is the best name for the AST argument (since it's not the Prism AST). The keyword argument name could be `ast`, which is simpler, but naming is difficult and I'm not confident that it's better 😅

Here is a proposed design for the `ast` parameter, assuming future RuboCop can use Prism AST:

- Current situation: When the `ast: ast` parameter is passed along with either `parser_engine: :parser_whitequark` or `parser_engine: parser_prism`, the AST is processed as a whitequark AST.
- Future extension: When the `ast: ast` parameter and `parser_engine: :prism` are passed, the AST would be processed as a Prism AST (note that `parser_engine: prism` is not yet implemented).

This would potentially allow for a future interface where RuboCop can use a direct Prism AST without `Prism::Translation::Parser`, and handle any incompatibilities in ASTs.